### PR TITLE
Issue 30: trim trace test support to real harness usage

### LINKED ISSUE #30: chore(tests): consolidate #[allow(dead_code)] and #[allow(unused_imports)] in tests/support/trace_llm.rs
## Background

`tests/support/trace_llm.rs` carries seven separate lint-suppression attributes across the file (identified during PR #23 review, https://github.com/leynos/axinite/pull/23):

| Line | Attribute | Item |
|------|-----------|------|
| 24 | `#[allow(unused_imports)]` | `pub use ironclaw::llm::recording::{…}` re-export block |
| 68 | `#[allow(dead_code)]` | `LlmTrace::steps` field |
| 199 | `#[allow(dead_code)]` | `patch_json_value` private function |
| 223 | `#[allow(dead_code)]` | `LlmTrace::new` constructor |
| 267 | `#[allow(dead_code)]` | `LlmTrace::patch_path` method |
| 283 | `#[allow(dead_code)]` | `LlmTrace::playable_steps` method |
| 314 | `#[allow(dead_code)]` | `impl TraceLlm` block |

These exist because the support module is compiled into each of the nine test-harness binaries, and each binary only uses a subset of the re-exported items and helper methods.

## Proposed Change

1. **Unused imports**: Audit the `pub use` re-export block on line 25 — remove any type aliases that are not referenced by any harness. For types that _are_ used by at least one harness, the `#[allow(unused_imports)]` is unavoidable given the shared-module pattern; consider replacing the per-item attribute with a module-level `#![allow(unused_imports)]` or switching to `#[allow(unused)]` at the relevant import level.
2. **Dead code**: For each method/function/field flagged above, verify it is genuinely required by at least one harness. If it is, consolidate all per-item `#[allow(dead_code)]` into a single module-level `#![allow(dead_code)]` at the top of `trace_llm.rs`. If an item is truly unused, remove it.

## Files Affected

- `tests/support/trace_llm.rs` — lines 24, 68, 199, 223, 267, 283, 314

## References

- PR #23: https://github.com/leynos/axinite/pull/23

### PR DESCRIPTION
**Pull request by leynos**
## Summary

- replace the global `tests/support/mod.rs` graph with harness-specific support roots
- remove the `tests/support/trace_llm.rs` facade and import trace helpers from their real owners
- keep trace helper APIs alive through real test coverage instead of compile-time touch references or wider lint suppression

## Validation

- `cargo test --tests --no-run`
- `make all`

Closes #30

## Summary by Sourcery

Restructure test support modules around harness-specific roots, removing the legacy global support graph and façade types while keeping trace helpers exercised via real tests.

New Features:
- Introduce dedicated support roots for e2e traces, support-unit tests, tools/trace-format tests, webhook tests, channel tests, and infrastructure tests.

Bug Fixes:
- Ensure directory setup helpers used by tests actually create target and suffixed directories, backed by new unit tests.

Enhancements:
- Split `LlmTrace` helper functionality into focused extensions for building, runtime loading, path mutation, and recorded-trace handling, scoped per harness.
- Expose limited `TraceLlm` internals and add optional diagnostics helpers so tests can inspect call counts and hint mismatches without relying on a broad façade.
- Refine webhook helper modules to separate pure HTTP routing utilities from webhook server lifecycle management used by specific tests.
- Tighten imports in tests to use trace and recording types from their real owners instead of a monolithic `trace_llm` façade, and remove unused helpers like the generic recorded-trace runner re-export.

Tests:
- Extend test-rig and cleanup support-unit tests to cover additional behaviors such as status-event clearing, tool event clearing, safety warnings, cost estimation, and directory-creation guarantees.

## Security Impact

No security impact. This PR restructures test-support modules only; no
production code paths, authentication logic, authorisation rules, or external
interfaces are modified.

## Database Impact

No database impact. No schema migrations, query changes, or data model
modifications are introduced.

## Blast Radius

Restricted to the test harness: `tests/support/` and `tests/support_unit_tests/`.
No production modules (`src/`) are changed. A regression introduced here would
cause test compilation or test-run failures only; it cannot affect deployed
behaviour.

## Rollback Plan

Revert the PR branch. The previous `tests/support/mod.rs` is preserved in git
history and can be restored without any migration step. No database or
infrastructure state is touched.

## Review Track

B — Refactor (test infrastructure only; no production behaviour change).

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | PR introduces significant new module structures with public helper methods lacking dedicated unit tests; all new functionality must be guarded by unit and behavioural tests. | Create dedicated unit test files for each newly introduced support module and convert .unwrap() calls to .expect() in cleanup\_tests.rs. |
| Developer Documentation | ⚠️ Warning | PR restructures test-support architecture with no corresponding developer documentation updates despite significant harness topology changes. | Update docs/developers-guide.md to document harness-specific support roots, import ownership boundaries, and rationale for splitting the shared support module. |

File: tests/support/trace_provider.rs [hunk 1]
```diff
@@ -24,16 +23,16 @@ use super::trace_types::LlmTrace;
...
     captured_requests: Vec<Vec<ChatMessage>>,
 }
 
 pub struct TraceLlm {
     model_name: String,
     steps: Vec<TraceStep>,
-    inner: Mutex<TraceLlmState>,
-    hint_mismatches: AtomicUsize,
+    pub(super) inner: Mutex<TraceLlmState>,
+    pub(super) hint_mismatches: AtomicUsize,
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_provider.rs:66]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/infrastructure.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 //! Infrastructure integration tests covering heartbeat, pairing, provider
 //! chaos, SIGHUP reload, and workspace functionality.
 
+#[path = "support/infrastructure.rs"]
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/infrastructure.rs:4]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/support_unit_tests.rs [hunk 3]
```diff
@@ -4,6 +4,7 @@
 //! and run exactly once, rather than being duplicated across every `e2e_*.rs`
 //! test binary that declares `mod support;`.
 
+#[path = "support/support_unit.rs"]
 mod support;
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support_unit_tests.rs:8]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/support/channels.rs [hunk 4]
```diff
@@ -0,0 +1,3 @@
+//! Shared support for channel-focused integration tests.
+
+pub mod telegram;
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/channels.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/support/trace_provider.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_provider.rs:66]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/developers-guide.md [hunk 5]
```diff
@@ -280,6 +279,124 @@ Keep broad behavioural coverage in the E2E trace fixtures, and place focused
...
+
+- declare `#[path = "support/<domain>.rs"] mod support;` in the harness
+  entrypoint;
+- add helper modules to the narrowest support root that owns the behaviour;
+- import from the real owner, for example `crate::support::trace_types::LlmTrace`
+  or `crate::support::trace_provider::TraceLlm`;
+- import recording data types directly from `ironclaw::llm::recording`;
+- do not recreate a global `tests/support/mod.rs` or a broad re-export facade.
+
+### Import ownership boundaries
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tests/support_unit_tests/trace_support_module_tests.rs](https://github.com/leynos/axinite/pull/165/files/57679717e5eca2c776166f652dd15b8817a42c69#diff-789926d7b61a355db9a2ca168a7cf67676da0a56a3015eaf4fbf1a185fe8cf1a)

Comment on lines +131 to +143

~~~rust
fn llm_trace_patch_path_rewrites_tool_call_arguments() {
    let from = "/tmp/original";
    let to = "/tmp/rewritten";
    let mut trace = LlmTrace::single_turn("patch-model", "patch file", vec![tool_call_step(from)]);

    let patched = trace.patch_path(from, to);

    assert_eq!(patched, 1);
    let TraceResponse::ToolCalls { tool_calls, .. } = &trace.turns[0].steps[0].response else {
        panic!("expected tool call step after patch");
    };
    assert_eq!(tool_calls[0].arguments["path"], serde_json::json!(to));
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=165&biomarker=Code+Duplication&filename=tests%2Fsupport_unit_tests%2Ftrace_support_module_tests.rs&method=)
The module contains 2 functions with similar structure: llm_trace_patch_path_ignores_empty_from_pattern,llm_trace_patch_path_rewrites_tool_call_arguments

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tests/support/trace_template_utils.rs](https://github.com/leynos/axinite/pull/165/files/dfceae3b557f238fbc059c13e51cdd6d4f1998cb#diff-193735e2bdddbf1299965c9489a86c268bc4d35fc15156663e948bea61b74125)

Comment on lines +18 to +40

~~~rust
pub(super) fn extract_tool_result_vars(messages: &[ChatMessage]) -> HashMap<String, String> {
    let mut vars = HashMap::new();
    for message in messages {
        if message.role != Role::Tool {
            continue;
        }
        let Some(call_id) = message.tool_call_id.as_deref() else {
            continue;
        };
        let content = unwrap_tool_output(&message.content);
        let Ok(json) = serde_json::from_str::<serde_json::Value>(&content) else {
            continue;
        };
        if let Some(obj) = json.as_object() {
            for (key, value) in obj {
                flatten_json_vars(&format!("{call_id}.{key}"), value, &mut vars);
            }
        } else {
            flatten_json_vars(call_id, &json, &mut vars);
        }
    }
    vars
}
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=165&biomarker=Bumpy+Road+Ahead&filename=tests%2Fsupport%2Ftrace_template_utils.rs&method=extract_tool_result_vars)
extract_tool_result_vars has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tests/support/trace_provider.rs (1)

`1-402`: _🛠️ Refactor suggestion_ | _🟠 Major_
**Split this module to satisfy the repository file-size limit.**
Reduce this file below the 400-line cap. Extract JSON/template substitution helpers (for example, Line 191-308) into a sibling support module and keep `TraceLlm` replay orchestration here.
As per coding guidelines, "Files must not exceed 400 lines in length".

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tests/support/assertions.rs (1)

`6-6`: _⚠️ Potential issue_ | _🟠 Major_
**Remove forbidden `#![allow(dead_code)]` suppression.**
This module-level lint suppression violates the coding guidelines, which state that `#[allow]` is forbidden and only narrowly scoped `#[expect(lint, reason = "...")]` is permitted as a last resort.
The PR's objective is to reduce lint suppressions by keeping helpers alive through actual test coverage. Either:
- Ensure all assertion helpers are exercised by tests (the preferred approach).
- If specific helpers are genuinely unused, remove them.
- If a helper must remain but is conditionally unused, apply a narrowly scoped `#[expect(dead_code, reason = "...")]` on that specific item with a clear justification.
As per coding guidelines: "Lints must not be silenced in Rust except as a last resort. `#[allow]` is forbidden. Only narrowly scoped `#[expect(lint, reason = \"...\")]` is allowed."

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

### ❌ Failed checks (3 errors, 1 warning, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | New tests validate field initialisation, fixture loading, path rewriting, filtering behaviour, and HTTP response codes. However, public functions extract\_tool\_result\_vars and substitute\_templates in tests/support/trace\_template\_utils.rs lack required Rustdoc documentation. | Add Rustdoc documentation (/// doc comments) to extract\_tool\_result\_vars and substitute\_templates, covering purpose, arguments, and behaviour per project standards. |
| Module-Level Documentation | ❌ Error | Functions extract\_tool\_result\_vars and substitute\_templates in tests/support/trace\_template\_utils.rs lack documentation comments, whilst other public functions in the PR correctly include them. | Add documentation comments to both extract\_tool\_result\_vars and substitute\_templates before their function signatures in trace\_template\_utils.rs. |
| Unit Architecture | ❌ Error | extract\_tool\_result\_vars() performs fallible JSON parsing but silently discards errors without exposing fallibility in its HashMap return type, violating the architectural requirement that fallible read operations must be explicit. | Refactor extract\_tool\_result\_vars() to return Result<HashMap<...>, JsonParseError>, propagating parse failures to callers rather than silently skipping invalid messages. |
| Concurrency And State | ⚠️ Warning | PR introduces concurrency-sensitive state management with proper primitives but lacks test coverage for failure paths, overflow scenarios, lock poisoning, and concurrent access patterns. | Add test coverage for next\_step overflow, document hint\_mismatches panic behaviour, test lock poisoning recovery, verify concurrent task cursor advancement monotonicity, and test counter overflow in setup\_test\_dir\_with\_suffix. |
| Testing (Property / Proof) | ❓ Inconclusive | PR introduces invariants (atomic counter overflow, replay cursor overflow, template expansion bounds) lacking property-based tests or formal verification coverage. | Implement property-based tests using proptest to validate invariants across arbitrary input spaces, concurrent access patterns, and state transitions. |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tests/support_unit_tests/cleanup_tests.rs (1)

`6-31`: _⚠️ Potential issue_ | _🟠 Major_
**Replace fixed `/tmp` test paths with per-test temporary paths.**
Use `tempfile`-scoped paths for these tests. Eliminate cross-process collisions and Unix-only path assumptions.

##### Patch

```diff
#[test]
 fn cleanup_guard_removes_file() {
-    let path = "/tmp/ironclaw_cleanup_guard_test.txt";
-    std::fs::write(path, "test").expect("should write test file");
+    let dir = tempfile::tempdir().expect("should create temp dir");
+    let path = dir.path().join("ironclaw_cleanup_guard_test.txt");
+    std::fs::write(&path, "test").expect("should write test file");
+    let path_str = path.to_string_lossy().into_owned();
     {
-        let _guard = CleanupGuard::new().file(path);
-        assert!(std::path::Path::new(path).exists());
+        let _guard = CleanupGuard::new().file(path_str.clone());
+        assert!(path.exists());
     }
-    assert!(!std::path::Path::new(path).exists());
+    assert!(!path.exists());
 }
```

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tests/support_unit_tests/cleanup_tests.rs](https://github.com/leynos/axinite/pull/165/files/593d4be402cebcb1898366e88417980166edc6b7#diff-e86ff5705e18fcdd7ed2aa06dc013cd002883c145538a803060ba8f194bc8dfd)

Comment on lines +13 to +15

~~~rust
    let file = tempfile::NamedTempFile::new().expect("should create temp file");
    let path = file.path().to_string_lossy().into_owned();
    std::fs::write(&path, "test").expect("should write test file");
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=165&biomarker=Code+Duplication&filename=tests%2Fsupport_unit_tests%2Fcleanup_tests.rs&method=)
The module contains 2 functions with similar structure: cleanup_guard_removes_dir,cleanup_guard_removes_file

File: tests/support_unit_tests/cleanup_tests.rs [hunk 6]
```diff
@@ -1,40 +1,108 @@
...
+    let counter = AtomicUsize::new(usize::MAX);
+    let result = counter.fetch_update(Ordering::Relaxed, Ordering::Relaxed, |current| {
+        current.checked_add(1)
+    });
+
+    assert!(
+        result.is_err(),
+        "fetch_update should fail when counter is at usize::MAX"
+    );
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support_unit_tests/cleanup_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

### ❌ Failed checks (2 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Critical test files referenced in the PR summary (trace\_provider\_tests.rs with 116 lines, property\_tests.rs with 72 lines) cannot be located in the repository. Review comments flag missing poisoned-lock regression tests and unimplemented chained-template expansion tests. | Locate and examine all test files in the PR commit history. Verify trace\_provider\_tests.rs poisoned-lock test coverage (lines 46–116) and trace\_template\_utils.rs chained-template test case {{a}} → "{{b}}" → 1 exist and would fail with incorrect implementations. |
| Unit Architecture | ❌ Error | TraceLlm::lock\_inner() provides Result-based error handling, but poisoned-lock regression tests are absent, template substitution early-returns break chained expansion, and test setup inconsistently uses .expect() instead of lock\_inner(). | Add poisoned-lock regression tests, remove early return in substitute\_templates to allow iterative resolution, update test setup to use lock\_inner() for consistency with production error handling. |
| Description check | ⚠️ Warning | The description covers main objectives, validation steps, and linked issue; however, it lacks explicit Security Impact, Database Impact, Blast Radius, and Rollback Plan sections required by the template. | Add missing template sections (Security Impact, Database Impact, Blast Radius, Rollback Plan) and assign Review track (recommend B for refactor). |
| Testing (Unit) | ⚠️ Warning | Pull request unit tests lack critical edge cases: missing lock-poison regression test in trace\_provider\_tests.rs, no chained-template test in trace\_template\_utils.rs, and substitute\_templates implementation prevents recursive expansion. | Add lock-poison regression test to trace\_provider\_tests.rs; fix substitute\_templates to permit whole-node string resolutions to continue into iterative expansion; add chained-template regression test. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces critical invariants (template termination, directory uniqueness, cursor overflow safety, lock poisoning, hint-mismatch bounds) but review comments explicitly identify two missing unit tests: poisoned-mutex regression test for trace\_provider.rs and chained-template-expansion regression test for trace\_template\_utils.rs. | Implement poisoned-lock test asserting LlmError::RequestFailed with 'TraceLlm state lock poisoned' message; add chained-expansion test validating {{a}}→'{{b}}'→1 fully resolves through iterations rather than returning on intermediate strings. |
| Performance And Resource Use | ⚠️ Warning | Code exhibits quadratic memory overhead via repeated HashSet cloning, early returns preventing chained template expansion, and avoidable allocations in loops. | Replace HashSet cycle detection with Option tracking previous result; fix early returns to permit iterative expansion; eliminate format!() allocations in loops using String mutations; refactor cloning in captured\_requests(). |
| Concurrency And State | ⚠️ Warning | PR omits required regression tests for poisoned-lock error path and chained template expansion, and implements early return preventing cascading template resolution. | Add poisoned-lock test to verify LlmError::RequestFailed, add chained-template expansion test, and modify substitute\_templates() to continue iterating on string-valued whole-string resolutions. |

File: tests/support/trace_provider.rs [hunk 7]
```diff
@@ -171,124 +191,32 @@ impl TraceLlm {
...
-                    if let Some(resolved) = vars.get(key) {
-                        *s = resolved.clone();
-                        return;
-                    }
-                }
+// The `hint_mismatches` overflow panic and `inner` lock-poison recovery path
+// are intentionally documented but not directly unit-tested here. Reaching the
+// overflow requires `usize::MAX` mismatches, and poisoning `inner` would require
+// a test-only mutation hook for `TraceLlm` internals that normal replay code
+// never needs.
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_provider.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/support/trace_provider.rs [hunk 8]
```diff
@@ -382,3 +310,99 @@ impl ironclaw::llm::NativeLlmProvider for TraceLlm {
...
+        assert_eq!(successes, 8);
+        let final_cursor = llm
+            .inner
+            .lock()
+            .expect("TraceLlm state lock should open")
+            .index;
+        assert_eq!(final_cursor, successes);
+        assert_eq!(llm.hint_mismatches.load(Ordering::SeqCst), 0);
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_provider.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/e2e_traces/recorded_trace.rs [hunk 9]
```diff
@@ -3,7 +3,7 @@
 //! Each test replays a recorded fixture through the full agent loop, verifying
 //! declarative `expects` from the JSON and any additional manual checks.
 
-use crate::support::test_rig::run_recorded_trace;
+use crate::support::run_recorded_trace;
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/e2e_traces/recorded_trace.rs:7]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/support/trace_template_utils.rs [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+        tool_message_fixture.tool_call_id = Some("call_bad".to_string());
+        let messages = [tool_message_fixture];
+
+        let result = extract_tool_result_vars(&messages);
+
+        assert!(result.is_err());
+        let err = result.expect_err("invalid JSON should return parse error");
+        assert_eq!(err.call_id, "call_bad");
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_template_utils.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tests/support/recorded_trace_runner.rs (1)

`12-24`: _⚠️ Potential issue_ | _🟠 Major_
**Stop panicking in shared support helpers; propagate errors instead.**
Replace panic/`expect` handling in `run_recorded_trace` with `Result` propagation and contextual errors. Keep test failures explicit without abort-style control flow from helper internals.

##### Patch

```diff
+use anyhow::{Context, Result};
@@
 #[cfg(feature = "libsql")]
-pub async fn run_recorded_trace(filename: &str) {
+pub async fn run_recorded_trace(filename: &str) -> Result<()> {
     let path = format!(
         "{}/tests/fixtures/llm_traces/recorded/{filename}",
         env!("CARGO_MANIFEST_DIR")
     );
     let trace = LlmTrace::from_file_async(&path)
         .await
-        .unwrap_or_else(|error| panic!("failed to load trace {filename}: {error}"));
+        .with_context(|| format!("failed to load trace {filename}"))?;
     let rig = TestRigBuilder::new()
         .with_trace(trace.clone())
         .build()
         .await
-        .expect("failed to build test rig");
+        .context("failed to build test rig")?;
     rig.run_and_verify_trace(&trace, Duration::from_secs(30))
         .await;
     rig.shutdown();
+    Ok(())
 }
```

As per coding guidelines, "In Rust production code and shared fixtures, avoid `.expect()` and propagate errors with `Result` and `?`".

##### 🤖 Prompt for AI Agents

#### tests/support/trace_types.rs (1)

`172-183`: _⚠️ Potential issue_ | _🟠 Major_
**Add diagnostic context to each fallible step in `load_trace_with_mutation`.**
Attach `.context(...)` to file read and JSON deserialisation paths so failures identify the exact failing stage.

##### 🩹 Proposed fix

```diff
+use anyhow::Context;
@@
 pub async fn load_trace_with_mutation<F>(
     path: impl AsRef<std::path::Path>,
     mutate: F,
 ) -> anyhow::Result<LlmTrace>
 where
     F: FnOnce(&mut serde_json::Value),
 {
-    let contents = tokio::fs::read_to_string(path).await?;
-    let mut value: serde_json::Value = serde_json::from_str(&contents)?;
+    let path_ref = path.as_ref();
+    let contents = tokio::fs::read_to_string(path_ref)
+        .await
+        .with_context(|| format!("failed to read trace JSON from {}", path_ref.display()))?;
+    let mut value: serde_json::Value = serde_json::from_str(&contents)
+        .context("failed to parse trace JSON into serde_json::Value")?;
     mutate(&mut value);
-    let trace = serde_json::from_value(value)?;
+    let trace = serde_json::from_value(value)
+        .context("failed to deserialize trace JSON into LlmTrace")?;
     Ok(trace)
 }
```

As per coding guidelines, "Use `anyhow` or `eyre` with `.context(...)` in Rust to preserve backtraces and add diagnostic context".

##### 🤖 Prompt for AI Agents

File: tests/support/trace_provider_tests.rs [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            successes += 1;
+        }
+    }
+
+    assert_eq!(successes, 8);
+    let final_cursor = llm
+        .inner
+        .lock()
+        .expect("TraceLlm state lock should open")
+        .index;
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_provider_tests.rs:150]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

File: tests/support/trace_provider_tests.rs [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        matches!(completion_err, LlmError::RequestFailed { .. }),
+        "expected request failure, got {completion_err:?}"
+    );
+    assert!(
+        completion_err
+            .to_string()
+            .contains("TraceLlm state lock poisoned"),
+        "expected poisoned-lock diagnostic, got {completion_err}"
+    );
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/trace_provider_tests.rs:92]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/developers-guide.md [hunk 13]
```diff
@@ -334,6 +334,95 @@ Keep broad behavioural coverage in the E2E trace fixtures, and place focused
...
+When migrating or adding tests:
+
+- declare `#[path = "support/<domain>.rs"] mod support;` in the harness
+  entrypoint;
+- add helper modules to the narrowest support root that owns the behaviour;
+- import from the real owner, for example `crate::support::trace_types::LlmTrace`
+  or `crate::support::trace_provider::TraceLlm`;
+- import recording data types directly from `ironclaw::llm::recording`;
+- do not recreate a global `tests/support/mod.rs` or a broad re-export facade.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/developers-guide.md [hunk 14]
```diff
@@ -334,6 +334,95 @@ Keep broad behavioural coverage in the E2E trace fixtures, and place focused
 helper regression tests in `tests/support_unit_tests/assertions_tests.rs` so
 they run once instead of being duplicated across every E2E binary.
 
+#### Integration test support roots
+
+Each top-level Rust file under `tests/` is a separate integration-test crate.
+The crate declares its own support boundary with a `#[path = "support/*.rs"]`
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

File: tests/support/test_rig/helpers.rs [hunk 15]
```diff
@@ -13,20 +15,21 @@ use super::TestRigBuilder;
...
         .with_trace(trace.clone())
         .build()
         .await
-        .expect("failed to build test rig");
+        .with_context(|| format!("building test rig for recorded trace: {filename}"))?;
     rig.run_and_verify_trace(&trace, Duration::from_secs(30))
         .await;
     rig.shutdown();
+    Ok(())
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/test_rig/helpers.rs:35]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
